### ISSUE TITLE
API is not able to handle hashers different to H256.

### ISSUE THREAD

**Issue by hbulgarini** [polkadot-js/api#3202]
Hey , when we try to access to a block information on a chain that contains a H512 hasher type: 
 Network -> Explorer -> Chain Info -> any block

i get  the following error:
```
Unable to retrieve the specified block details. Unable to retrieve header and parent from supplied hash
```

Please take into account that we are adding already the custom type:
`  "BlockHash": "H512"`

Doing a research on the api code we detected that the hash is being hardcoded to H256:
https://github.com/polkadot-js/api/blob/d2ff350db963600fb5256152ef6946856e1bbb96/packages/types/src/types/registry.ts#L112 
https://github.com/polkadot-js/api/blob/d2ff350db963600fb5256152ef6946856e1bbb96/packages/types/src/codec/Struct.ts#L228

Here is the link to our custom chain:
[Millau](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.millau.brucke.link&types=eNp9VVlPGzEQ%2Fiton0Dah3IEUYqQuCqiAkEE0YcKrcx6snFx7K0P2jTiv3dm7T1D%2BmbPd3i8M%2BNdJWecG7A2OU7O8lx75cY8SZMbrV99OdXe5DCAboWUzJ9Lnb9eMztH9Hq0u9cH7vziBQxC%2FvAAkQfBpNM9yd7osA%2B0kn0yC8g1ME7BVVIyA8plUT40TBNVyy%2F0omS5O1mzPkWWdcxBZrR2H5rAH2eEsiK3mzlcFGBb6DJs3%2BuM4x4zlrqgr%2FoEecwlQGMHi9Mhn4KkyQDvQYuJm1e3IbU%2FotTPvJtrI5wAezFnqqCyEFrHl2NOtIDZRyPgYdMVpsDk0yeE4gJVWllQ1lPC7TowIw939wYesAvEgs7ubN7xNt3iDjskok2F1xsFGUbwAnjXZj3xLqlxW%2B%2Bhhli3T7%2Bb0mS8KLVxLb5K5jWzr0wTA7%2B8wAHJfnrrxEzkzAmtkPiiNX0TYbOZUEyKv8DbqBUFxrJ5uMikJM3J8IpVFzTlm0LVNawtM81djd4IbCr0BZcJOgfZOIzURSDZEgxu%2BlNaBX74oy9bB8%2B0Vdj6KocqHCacqc7uFp8AVsCdRk4ztjFYsbbHPN3C8E4b%2FwbLqs%2FRKCQVLXEYyeaYKt21xWzH6gUz5DHpYcoBJJdL5hh5m0Cs52i765du9XfNl9ihOZDMuizXaibMAnhWZfRBQhPv1g6V%2BEaguAAFBpcbxHRExTOQg3jbTNOSE82r0nj1n1Tivs6iZEupGX3WCNzHQJrMgPTnDL98T1ozus9GfcFI%2BVpJQ3lrqJGtkpxJmvdWjs1XCGr3C0QmYYNtWEKevWFdwiiEF%2Fs3iGJOD873sMC8OqLu4xZ%2BK%2FF52qaOemSmABebgYLUufv7z1hjL52Y4jiBSZs1c94A6YJTq4uy6kXqKHunX%2FG90Wj3c%2FsTmppB4CrnFmvwsVt19ibD8CPsGIZA1%2FBwRIbv%2FwByM4Kb)

So is there any overwrite we can do for changing the default hasher to H512?
@tomusdrw

Thanks!

**Comment by tomusdrw** [comment]
> So is there any overwrite we can do for changing the default hasher to H512?

Actually, the hasher we use is a custom hasher (`blake2(content) ++ keccak256(content)`), so we would need a way to overwrite the hasher (globally) and produce `H512` outputs.

**Comment by jacogr** [comment]
There are 2 parts to hashing -

- the actual hashing taking place
- the actual type that is being created from that output

You are referring to the second part above. The registry allows you to configure the type of hashing for the chain - https://github.com/polkadot-js/api/blob/master/packages/types/src/create/registry.ts#L371-L373

Which is employed here in all codec hashers - https://github.com/polkadot-js/api/blob/master/packages/types/src/create/registry.ts#L318-L320 (This is where the raw Uint8array is converted to the defined hash type)

So basically, you need a `CodecHash` typ[e to be added that gets used above. (And then inject the hasher as to the chain specific - ideally, this is probably part of the API config, so it gets applies to all registry instances as created)

**Comment by hbulgarini** [comment]
thank you jaco!

two questions:
1) Is there any example where i can see the use of `api.resgistry.setHasher` ?
2) Is there any way to set the hasher in the apps portal?

**Comment by jacogr** [comment]
1. No, but it is just `registry.setHasher` which takes a function that accepts a `Uint8Array` as input and outputs a `Uint8Array`. This would work swimmingly with 32-byte outputs, in your case it won't since it converts the output into a H256 value.

2. The apps UI doesn't. It can if the config is part of the typesBundle and the apps-config specifies it for a specific chain. (Which is certainly advisable for this type of thing so all config is in a single place)
